hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/ImmutableSegment.java

```diff
@@ -43,4 +112,144 @@ public KeyValueScanner getKeyValueScanner() {
     return new CollectionBackedScanner(getCellSet(), getComparator());
   }
 
+  /**------------------------------------------------------------------------
+   * Change the CellSet of this ImmutableSegment from one based on ConcurrentSkipListMap to one
+   * based on CellArrayMap.
+   * If this ImmutableSegment is not based on ConcurrentSkipListMap , this is NOP
+   * For now the change from ConcurrentSkipListMap to CellChunkMap is not supported, because
+   * this requires the Cell to know on which Chunk it is placed.
+   *
+   * Synchronization of the CellSet replacement:
+   * The reference to the CellSet is AtomicReference and is updated only when ImmutableSegment
+   * is constructed (single thread) or flattened. The flattening happens as part of a single
+   * thread of compaction, but to be on the safe side the initial CellSet is locally saved
+   * before the flattening and then replaced using CAS instruction.
+   */
+  public boolean flatten() {
+    if (isFlat) return false;
+    CellSet oldCellSet = getCellSet();
+    int numOfCells = getCellsCount();
+
+    // each Cell is now represented in either in CellArrayMap or in CellChunkMap
+    constantCellMetaDataSize = ClassSize.CELL_ARRAY_MAP_ENTRY;
+
+    // arrange the meta-data size, decrease all meta-data sizes related to SkipList
+    incSize(
+        -(ClassSize.CONCURRENT_SKIPLISTMAP + numOfCells * ClassSize.CONCURRENT_SKIPLISTMAP_ENTRY));
+    // add size of CellArrayMap and meta-data overhead per Cell
+    incSize(ClassSize.CELL_CHUNK_MAP + numOfCells * constantCellMetaDataSize);
+
+    CellSet  newCellSet = recreateCellArrayMapSet(numOfCells); // build the CellSet CellArrayMap based
+    setCellSet(oldCellSet,newCellSet);
+    return true;
+  }
+
+  /////////////////////  PRIVATE METHODS  /////////////////////
+  /*------------------------------------------------------------------------*/
+  // Create CellSet based on CellArrayMap from compacting iterator
+  private CellSet createCellArrayMapSet(int numOfCells, MemStoreCompactorIterator iterator) {
```

Comment:
Okay, getting the idea now. Please don't get mad, but you might need one more step, maybe not in this patch. 

In this implementation, ImmutableSegment is actually a factory. Most of the logic is around creation/transformation of CellSet's. It makes sense to me to subclass ImmutableSegment into the following classes: (1) ISWrapper, (2) ISCellObjectArray, and (3) ISChunkArray. Then, put a factory on top of them - to be invoked by the Compactor. 
Take a look at the current code - there is near-zero overlap between these code paths - calling for polymorphism.  

At least, please document this design point.